include/zephyr/net/coap_client.h

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/net/coap.h>
+
+#define COAP_VERSION 1
+#define COAP_PATH_ELEM_DELIM '/'
+#define COAP_PATH_ELEM_QUERY '?'
+#define MAX_COAP_MSG_LEN CONFIG_COAP_ASYNC_CLIENT_BUFFER_SIZE
+#define COAP_POOL_SLEEP 500
+#define COAP_SEPARATE_TIMEOUT 6000
+#define DEFAULT_RETRY_AMOUNT 5
+
+/**
+ * @typedef coap_client_response_cb_t
+ * @brief Callback for CoAP request.
+ *
+ * This callback is called for responses to CoAP client requests.
+ * It is used to indicate errors, response codes from server or to deliver payload.
+ * Blockwise transfers cause this callback to be called sequentially with increasing payload offset
+ * and only partial content in buffer pointed by payload parameter.
+ *
+ * @param result_code Result code of the response. Negative if there was a failure in send.
+ *                    @ref coap_response_code for positive.
+ * @param offset Payload offset.
+ * @param payload Buffer containing the payload from the response. NULL for empty payload.
+ * @param len Size of the payload.
+ * @param user_data User provided context.
+ * @param last_block Indicates the last block of the response.
+ */
+typedef void (*coap_client_response_cb_t)(uint8_t result_code,
+					  size_t offset, size_t len, const uint8_t *payload,
+					  bool last_block, void *user_data);
+
+/** @cond INTERNAL_HIDDEN */
+struct coap_client_request {
+	enum coap_method method;      /**< Method of the request */
+	bool confirmable;	      /**< CoAP Confirmable/Non-confirmable message */
+	const char *path;	      /**< Path of the requested resource */
+	enum coap_content_format fmt; /**< Content format to be used */
+	uint8_t *payload;	      /**< User allocated buffer for send request */
+	size_t len;		      /**< Length of the payload */
+	coap_client_response_cb_t cb; /**< Callback when response received */
+	void *user_data;	      /**< User provided context */
+};
+/** @endcond */
+
+/**
+ * @brief Internal structure of CoAP client, shouldn't be modified.
+ */
+struct coap_client {
```

Comment:
The current implementation has a stack per CoAP client, if I were to do multiple requests at the same time, I either need to do these sequentially or create a client for each request, which is very costly.
This becomes even more apparent if I need to observe multiple server resources (long lasting requests).

It would be great if a single client could support multiple requests.
I think this is doable if you would move some members from `struct coap_client` to `struct coap_client_req` to keep track if a single request state. And let the client keep track of al ongoing requests (for example in a linked list).

Reply:
The multiple request is planned also, and doesn't need changes to the public API.